src/com/owncloud/android/ui/fragment/OCFileListFragment.java

```diff
@@ -154,8 +177,170 @@ public void onActivityCreated(Bundle savedInstanceState) {
         setListAdapter(mAdapter);
 
         registerLongClickListener();
+
+        boolean hideFab = (args != null) && args.getBoolean(ARG_HIDE_FAB, false);
+        if (hideFab) {
+            setFabEnabled(false);
+        } else {
+            setFabEnabled(true);
+            registerFabListeners();
+
+            // detect if a mini FAB has ever been clicked
+            final SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(getActivity());
+            if(prefs.getLong(KEY_FAB_EVER_CLICKED, 0) > 0) {
+                miniFabClicked = true;
+            }
+
+            // add labels to the min FABs when none of them has ever been clicked on
+            if(!miniFabClicked) {
+                setFabLabels();
+            } else {
+                removeFabLabels();
+            }
+        }
   }
 
+    /**
+     * adds labels to all mini FABs.
+     */
+    private void setFabLabels() {
+        getFabUpload().setTitle(getResources().getString(R.string.actionbar_upload));
+        getFabMkdir().setTitle(getResources().getString(R.string.actionbar_mkdir));
+        getFabUploadFromApp().setTitle(getResources().getString(R.string.actionbar_upload_from_apps));
+    }
+
+    /**
+     * registers all listeners on all mini FABs.
+     */
+    private void registerFabListeners() {
+        registerFabUploadListeners();
+        registerFabMkDirListeners();
+        registerFabUploadFromAppListeners();
+    }
+
+    /**
+     * registers {@link android.view.View.OnClickListener} and {@link android.view.View.OnLongClickListener}
+     * on the Upload mini FAB for the linked action and {@link Toast} showing the underlying action.
+     */
+    private void registerFabUploadListeners() {
+        getFabUpload().setOnClickListener(new View.OnClickListener() {
+            @Override
+            public void onClick(View v) {
+                Intent action = new Intent(getActivity(), UploadFilesActivity.class);
+                action.putExtra(
+                        UploadFilesActivity.EXTRA_ACCOUNT,
+                        ((FileActivity)getActivity()).getAccount()
+                );
+                //startActivityForResult(action, ACTION_SELECT_MULTIPLE_FILES);
+                // this flow seems broken;
+                getActivity().startActivityForResult(
```

Comment:
As mentioned in the other comment, this is now exclusively handled by the fragment, since the activity doesn't trigger these actions anymore, I slightly refactored the code
* so I removed the method in the activity https://github.com/owncloud/android/commit/8d038e89a9560e190f282d5b074b909193292822
* moved the requestCode to the fragment https://github.com/owncloud/android/commit/f34a9ddf5b2c2a5bb326fc08ddac65c3e3546737
* also removed the "other" action implementations since they are now handeled by the fab implementation in the fragment https://github.com/owncloud/android/commit/7e82baa8bc068bee956f6f03b61c06151c953eac
* shortened the upload call by creating a static helper method within the upload activity https://github.com/owncloud/android/commit/7e82baa8bc068bee956f6f03b61c06151c953eac